File: ch19p645unwind/ch19p645unwind/ViewController3.m

#import "ViewController3.h"

@interface ViewController3 ()

@end

@implementation ViewController3

-(BOOL)respondsToSelector:(SEL)aSelector {
    if ([NSStringFromSelector(aSelector) isEqualToString:@"unwind:"]) {
        NSLog(@"%@ is asked if it responds to %@", self, NSStringFromSelector(aSelector));
    }
    BOOL result = [super respondsToSelector:aSelector];
    return result;
}

-(UIViewController *)viewControllerForUnwindSegueAction:(SEL)action fromViewController:(UIViewController *)fromViewController withSender:(id)sender {
    UIViewController* vc = [super viewControllerForUnwindSegueAction:action fromViewController:fromViewController withSender:sender];
    NSLog(@"%@ returns %@ from vc for unwind segue", self, vc);
    return vc;
}

-(UIStoryboardSegue *)segueForUnwindingToViewController:(UIViewController *)toViewController fromViewController:(UIViewController *)fromViewController identifier:(NSString *)identifier {
    NSLog(@"%@ was asked for segue", self);
    return nil;
}


-(IBAction)unwind: (UIStoryboardSegue*) seg {
    NSLog(@"%@", @"view controller 3 unwind is never called");
}

-(BOOL)canPerformUnwindSegueAction:(SEL)action fromViewController:(UIViewController *)fromViewController withSender:(id)sender {
    NSLog(@"%@", @"view controller 3 can perform is never called");
    return NO;
}

-(BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender {
    NSLog(@"%@", @"view controller 3 should perform returns YES");
    return YES;
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    NSLog(@"%@", @"view controller 3 prepare for segue is called");
}


@end

File: bk2ch06p327unwind/ch19p645unwind/ViewController3.swift

import UIKit

class ViewController3 : UIViewController {
    
    override func respondsToSelector(aSelector: Selector) -> Bool {
        if NSStringFromSelector(aSelector) == "unwind:" {
            print("\(self) is asked if it responds to \(aSelector)")
        }
        let result = super.respondsToSelector(aSelector)
        return result
    }

    override func allowedChildViewControllersForUnwindingFromSource(source: UIStoryboardUnwindSegueSource) -> [UIViewController] {
        print("view controller 3 allowed child vcs called...")
        let vcs = super.allowedChildViewControllersForUnwindingFromSource(source)
        print("view controller 3 the source was \(source.sourceViewController) sent by \(source.sender)")
        print("view controller 3 returns \(vcs) from allowed child vcs")
        return vcs
    }
    
    override func unwindForSegue(unwindSegue: UIStoryboardSegue, towardsViewController subsequentVC: UIViewController) {
        print("view controller 3 was told to unwind for segue")
        super.unwindForSegue(unwindSegue, towardsViewController: subsequentVC)
    }

    
    @IBAction func unwind(seg:UIStoryboardSegue!) {
        fatalError("view controller 3 unwind should never be called")
    }
    
    override func canPerformUnwindSegueAction(action: Selector, fromViewController: UIViewController, withSender sender: AnyObject) -> Bool {
        let result = super.canPerformUnwindSegueAction(action, fromViewController: fromViewController, withSender: sender)
        print("view controller 3 can perform returns \(result)")
        return result
    }
    
    override func shouldPerformSegueWithIdentifier(identifier: String, sender: AnyObject?) -> Bool {
        let result = super.shouldPerformSegueWithIdentifier(identifier, sender: sender)
        print("view controller 3 should perform returns \(result)")
        return result
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject!) {
        print("view controller 3 prepare for segue is called")
    }

}
